### DISCUSSION TITLE
I added a -p option for poliqarp output

### DISCUSSION THREAD

**Discussion by rmast** [rmast/revealshapes#4] [Announcements]
@jsbien The changes needed for filename-manipulation urged me to require C++20 (or C++2a), but the executable from the Action-pipeline should nou accept a -p option to use your specified poliqarp-output. I tested the resulting highlights in a single page djvu, and they match symbols. The output is still to stdout.

**Comment by jsbien** [comment]
Great, but how to use it? I'm still unable to compile it on Debian.

**Comment by rmast** [comment]
I asked the DjVuLibre maintainers to provide the necessary headerfiles in the libdjvulibre-dev package, as those are a dependency. You can watch the github-action how to provide them with the current version.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by jsbien** [reply]
> You can watch the github-action how to provide them with the current version.

OK, but I have no experience with github-action. What do you exactly suggest?

djview4poliqarp and djview4shape contain `src/libdjvu` directory with various DjVu headers. Isn't this a way to circumvent the problem? I copied it to the `revealshapes` directory, but that is not sufficient, files like "DjVuDocument.h" are still not found despite the fact they are present in libdjvu.

**Comment by rmast** [comment]
Just pick the necessary steps out of

https://github.com/rmast/revealshapes/blob/main/.github/workflows/build-revealshapes-ubuntu-20.04.yml

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by rmast** [comment]
So:

sudo apt install libdjvulibre-dev

cd  ~/revealshapes

apt-get source libdjvulibre-dev

mkdir -p external
cp -r djvulibre-*/libdjvu/*.h external

./autogen.sh
make


Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jsbien** [comment]
Thanks. 

I get
`revealshapes.cpp: In function ‘int process_document(int, int, DJVU::GP<DJVU::DjVuDocument>)’:`
`revealshapes.cpp:73:56: error: cannot convert ‘std::__cxx11::basic_string<char>::iterator’ to ‘const char*’`
`   73 |         sorigString.erase(std::remove(sorigString.begin(), sorigString.end(), '\"'), sorigString.end());`
`      |                                       ~~~~~~~~~~~~~~~~~^~`
`      |                                                        |`
`      |                                                        std::__cxx11::basic_string<char>::iterator`

Perhaps I'm using too old or too new compiler? ` cpp --version` gives `cpp (Debian 12.2.0-14) 12.2.0`. This is the default version, but version 11 is still available. You mentioned C++20 (or C++2a) but I don't see anything like this on Debian.

Looks like I should use `-std=c++20` parameter and/or invoke the compiler with g++, not gcc, but I see no place in Makefile where I could make this changes.

**Comment by rmast** [comment]
The github actions use ubuntu 20.04 and 22.04.

I had to require language compatibility -std C++2a/20 for this language construct.
By default C++11 was selected. This setting should be in configure.ac or Makefile.am.

Mint 20.3 has C++ version 9.something as default, I also installed version 10, but both were compiling this.

C++ version 9 only supports language C++2a.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by jsbien** [reply]
It's in Makefile.am. I will look into it later.
Just curious, why ./configure is not used?

**Comment by rmast** [comment]
You can throw away configure as that is generated.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jsbien** [comment]
I used `remake`.

Looks like the compiler is called correctly
`g++ -DHAVE_CONFIG_H -I.  -Iexternal  -pthread -std=c++2a -g -O3 -Wall -DNDEBUG -MT revealshapes.o -MD -MP -MF $depbase.Tpo -c -o revealshapes.o revealshapes.cpp &&\`
but there is still an error. I enclose the full log.

Any ideas?
[badlog.zip](https://github.com/rmast/revealshapes/files/15078912/badlog.zip)

**Comment by rmast** [comment]
I guess the strictness of your new compiler prohibits an implicit conversion that was allowed on mine, despite the compatibility setting. I'll ask ChatGPT or alike for a solution.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by rmast** [comment]
ChatGPT solved it. You can pull the new commit and try.

**Reply by jsbien** [reply]
It works, thanks!

But the page numbers should start at 1, not 0.

When djview-based viewer gets a wrong page number it doesn't complain but open the page 1, that's why you didn't noticed it.

**Comment by jsbien** [comment]
![Screenshot from 2024-04-24 13-06-05](https://github.com/rmast/revealshapes/assets/8024702/3b51535c-1f60-49f2-8b06-06a7272ce0eb)
